mesonbuild/interpreterbase.py

```diff
@@ -572,68 +590,70 @@ def evaluate_comparison(self, node):
+        # but we actually guarantee earlier that both types are the same
         elif node.ctype == '<':
-            return val1 < val2
+            return val1 < val2   # type: ignore
         elif node.ctype == '<=':
-            return val1 <= val2
+            return val1 <= val2  # type: ignore
         elif node.ctype == '>':
-            return val1 > val2
+            return val1 > val2   # type: ignore
         elif node.ctype == '>=':
-            return val1 >= val2
+            return val1 >= val2  # type: ignore
         else:
             raise InvalidCode('You broke my compare eval.')
 
-    def evaluate_andstatement(self, cur):
+    def evaluate_andstatement(self, cur: mparser.AndNode) -> Union[bool, Disabler]:
         l = self.evaluate_statement(cur.left)
-        if is_disabler(l):
+        if isinstance(l, Disabler):
             return l
         if not isinstance(l, bool):
             raise InterpreterException('First argument to "and" is not a boolean.')
         if not l:
             return False
         r = self.evaluate_statement(cur.right)
-        if is_disabler(r):
+        if isinstance(r, Disabler):
             return r
         if not isinstance(r, bool):
             raise InterpreterException('Second argument to "and" is not a boolean.')
         return r
 
-    def evaluate_orstatement(self, cur):
+    def evaluate_orstatement(self, cur: mparser.OrNode) -> Union[bool, Disabler]:
         l = self.evaluate_statement(cur.left)
-        if is_disabler(l):
+        if isinstance(l, Disabler):
             return l
         if not isinstance(l, bool):
             raise InterpreterException('First argument to "or" is not a boolean.')
         if l:
             return True
         r = self.evaluate_statement(cur.right)
-        if is_disabler(r):
+        if isinstance(r, Disabler):
             return r
         if not isinstance(r, bool):
             raise InterpreterException('Second argument to "or" is not a boolean.')
         return r
 
-    def evaluate_uminusstatement(self, cur):
+    def evaluate_uminusstatement(self, cur) -> Union[int, Disabler]:
         v = self.evaluate_statement(cur.value)
-        if is_disabler(v):
+        if isinstance(v, Disabler):
             return v
         if not isinstance(v, int):
             raise InterpreterException('Argument to negation is not an integer.')
         return -v
 
     @FeatureNew('/ with string arguments', '0.49.0')
-    def evaluate_path_join(self, l, r):
+    def evaluate_path_join(self, l: str, r: str) -> str:
         if not isinstance(l, str):
             raise InvalidCode('The division operator can only append to a string.')
         if not isinstance(r, str):
             raise InvalidCode('The division operator can only append a string.')
         return self.join_path_strings((l, r))
 
-    def evaluate_division(self, l, r):
+    def evaluate_division(self, l: Any, r: Any) -> Union[int, str]:
```

Comment:
I think the way that you want to type this is:

```python

@typing.overload
def evalulate_division(self, l: str, r: str> -> str:
    pass

@typing.overload
def evaluate_division(self, l: int, r: int) -> int:
    pass

def evaluate_division(self, l, r):
    <implementation>
```

Not that this is the syntax to define highly constrained generic types, so you don't annotate the implementation, it assumes that you're doing the right stuff in the overloads to make this accurate.

Reply:
I think it should be `Any` in this case (rather than an overload) since `evalulate_division` itself does the type check. With an overload, I have to move the type checking to the caller, which basically makes `evaluate_division` redundant.